### PR TITLE
Timer refactor

### PR DESCRIPTION
**Pull request by jamesadevine**
Hi @finneyj @mmoskal 

I've added a low level timer interface (see [here](https://github.com/lancaster-university/codal-core/blob/master/inc/driver-models/LowLevelTimer.h)) to codal-core, and in this PR are two implementations of that interface for TC and TCC timers on the samd.

Could you let me know what you think of the driver model and let me know if I've missed anything?

I have only considered Timer/Counter scenarios in the interface and not things like PWM generation.

### CODE REVIEW

**Comment by jamesadevine** [comment]
Example app:

```cpp
#ifdef USE_TC
#include "SAMDTCTimer.h"
#pragma message "TC Timer"
SAMDTCTimer t(TC4, TC4_IRQn);
#else
#include "SAMDTCCTimer.h"
#pragma message "TCC Timer"
SAMDTCCTimer t(TCC0, TCC0_IRQn);
#endif

ZPin led(ID_PIN_LED, PIN_PA17, PIN_CAPABILITY_DIGITAL);
int state = 0;

void timerIrq(uint16_t channel)
{
    DMESG("I");
    led.setDigitalValue(state);
    t.offsetCompare(0,500000);
    state = !state;
}

int main()
{
    DMESG("HELLO");
    t.setIRQ(timerIrq);
    t.setCompare(0,500000);
    t.enable();
    DMESG("EN");

    while(1);
}
```

**Comment by jamesadevine** [comment]
- [ ] Regression test on samd51

File: src/SAMDTCCTimer.cpp [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+    return DEVICE_OK;
+}
+
+int SAMDTCCTimer::clearCompare(uint8_t channel)
+{
+    if (channel > getChannelCount())
+        return DEVICE_INVALID_PARAMETER;
+
+    // add channel to MC0, MC0 is 4, MC1 is 5
+    tcc->INTENCLR.reg = (1 << (TCC_INTENCLR_MC0_Pos + channel));
```
Comment (mmoskal) [inline src/SAMDTCCTimer.cpp]: Clear Mc as well? So later offset compare works

File: src/SAMDTCTimer.cpp [hunk 2]
```diff
@@ -0,0 +1,442 @@
...
+            while (tc->COUNT8.STATUS.bit.SYNCBUSY);
+#endif
+            // add channel to MC0, MC0 is 4, MC1 is 5
+            tc->COUNT8.INTENSET.reg = (1 << (TC_INTENSET_MC0_Pos + channel));
+            break;
+        case BitMode16:
+            tc->COUNT16.CC[channel].reg = value;
+#ifdef SAMD21
+            while (tc->COUNT16.STATUS.bit.SYNCBUSY);
+#endif
```
Comment (mmoskal) [inline src/SAMDTCTimer.cpp]: Maybe move these three wait functions to three macros and define them to nothing on 51?

File: src/SAMDTCCTimer.cpp (hunk 1 above)
Comment (jamesadevine) [reply src/SAMDTCCTimer.cpp]: Will do

File: src/SAMDTCTimer.cpp (hunk 2 above)
Comment (jamesadevine) [reply src/SAMDTCTimer.cpp]: Cheers, will do

**Comment by jamesadevine** [comment]
@mmoskal A timer still works on SAMD51 with the caveat that only 16-bit mode seems to work. 32 bit mode seems to not work and I can't determine why (nor do I want to invest the time).

**Comment by jamesadevine** [comment]
```
SAMDTCTimer             sysTimer(TC0, TC0_IRQn);
Timer                   timer(sysTimer);
```

**Pull request merged**